server/sonar-server/src/main/java/org/sonar/server/favorite/ws/SearchAction.java

```diff
@@ -86,83 +89,78 @@ private static SearchRequest toWsRequest(Request request) {
   private SearchResults toSearchResults(SearchRequest request) {
     userSession.checkLoggedIn();
     try (DbSession dbSession = dbClient.openSession(false)) {
-      SearchResults.Builder builder = SearchResults.newBuilder(request);
-      addAuthorizedProjectUuids(dbSession, builder);
-      addFavorites(builder);
-      return builder.build();
+      List<ComponentDto> authorizedFavorites = favoriteFinder.list().stream()
+        .filter(isAuthorized(dbSession))
+        .collect(Collectors.toList());
+      Paging paging = Paging.forPageIndex(request.getPage()).withPageSize(request.getPageSize()).andTotal(authorizedFavorites.size());
+      List<ComponentDto> displayedFavorites = authorizedFavorites.stream()
+        .skip(paging.offset())
+        .limit(paging.pageSize())
+        .collect(Collectors.toList());
+      Map<String, OrganizationDto> organizationsByUuid = getOrganizationsOfComponents(dbSession, displayedFavorites);
+      return new SearchResults(paging, displayedFavorites, organizationsByUuid);
     }
   }
 
-  private void addFavorites(SearchResults.Builder builder) {
-    builder.allFavorites = favoriteFinder.list();
+  private Predicate<ComponentDto> isAuthorized(DbSession dbSession) {
+    Collection<String> rootProjectsUuids = dbClient.authorizationDao().selectAuthorizedRootProjectsUuids(dbSession, userSession.getUserId(), UserRole.USER);
```

Comment:
The algorithm is to load all the projects visible by user (so potentially a huge number) and then to filter favorites ? If yes, then what about calling `userPermission#hasComponentPermission()` which contains a cache ?